src/attic/libauth-4.0-p5/authuser.3
.TH authuser 3
.SH NAME
authuser \- remote authentication library using the Authentication Server
.SH SYNTAX
.B #include <authuser.h>

unsigned short \fBauth_tcpport\fR;

char *\fBauth_xline(\fIuser,fd,&inremote\fB)\fR;

int \fBauth_fd(\fIfd,&inremote,&local,&remote\fB)\fR;
.br
int \fBauth_fd2(\fIfd,&inlocal,&inremote,&local,&remote\fB)\fR;

int \fBauth_tcpsock(\fIfd,inlocal,inremote\fB)\fR;

char *\fBauth_tcpuser(\fIinremote,local,remote\fB)\fR;
.br
char *\fBauth_tcpuser2(\fIinlocal,inremote,local,remote\fB)\fR;
.br
char *\fBauth_tcpuser3(\fIinlocal,inremote,local,remote,ctimeout\fB)\fR;
.br
char *\fBauth_tcpuser4(\fIinlocal,inremote,local,remote,ctimeout,rtimeout\fB)\fR;

char *\fBauth_sockuser(\fIs,local,remote\fB)\fR;

char *\fIuser\fP;
.br
int \fIfd\fP;
.br
int \fIs\fP;
.br
unsigned long \fIinlocal\fP;
.br
unsigned long \fIinremote\fP;
.br
unsigned short \fIlocal\fP;
.br
unsigned short \fIremote\fP;
.br
int \fItimeout\fP;
.SH DESCRIPTION
The
.B authuser
library provides a simple interface for
finding out the remote identity
of a connection through the
Authentication Server
as specified by RFC 931.
Use the -lauthuser loader option
to compile a program with this library.

.B auth_xline(\fIuser,fd,&inremote\fB)
returns a line of the form X-Auth-User: user or X-Forgery-By: username,
depending upon what the host on the other side of
.I fd
thinks of the user.
This is particularly appropriate for
mail and news headers.
.PP
If the remote host reports that
.I user
owns the connection on that side,
.B auth_xline
will return X-Auth-User: user.
If the remote host reports that a different
username
owns the connection,
.B auth_xline
will return X-Forgery-By: username.
If user is NULL,
it returns X-Auth-User: username
with the username reported by the remote host.
If 
.I fd
is not a TCP connection
or authentication is impossible,
.B auth_xline
returns NULL, setting errno appropriately.

The line is not cr-lf terminated.
It is stored in a static area
which is overwritten on each call to
.B auth_xline.
.B auth_xline
places the Internet address of the other host into
.I inremote.

.B auth_fd2(\fIfd,&inlocal,&inremote,&local,&remote\fB)
retrieves address information from the connection in socket
.I fd.
It places the
Internet addresses of the connection into
.I inlocal
and
.I inremote
and the local and remote
TCP ports into 
.I local
and 
.I remote.
.B auth_fd2
returns -1 upon error, setting errno appropriately.

.B auth_tcpuser2(\fIinlocal,inremote,local,remote\fB)
returns the name of the user on the other end of the TCP connection
between
.I remote@inremote
and
.I local@inlocal.
If authentication is impossible,
.B auth_tcpuser2
returns
NULL, setting errno appropriately.
The user name is stored in a static area
which is overwritten on each call to
.B auth_tcpuser2,
.B auth_tcpuser,
.B auth_sockuser,
and
.B auth_xline.

.B \fIs\fB = auth_tcpsock(\fIfd,inlocal,inremote\fB)
sets
.I s
to a non-blocking socket which is connecting to the Authentication
Server at
.I inremote.
It returns -1 on error, setting errno appropriately.
.B auth_sockuser(\fIs,local,remote\fB)
makes sure that the socket has connected and
then does the same job as
.B auth_tcpuser2,
returning the name of the user on the other end of the TCP connection
between
.I remote@inremote
and
.I local@inlocal,
or NULL (with errno set) if authentication is not possible.
.I s
is closed by
.B auth_sockuser.
The advantage of
using
.B auth_tcpsock
and
.B auth_sockuser
instead of
.B auth_tcpuser2
is that you can perform other actions while waiting for the
authentication request to complete.
You can select
.I s
for writing
to see if it is ready for
.B auth_sockuser
yet.

.B auth_tcpuser3(\fIinlocal,inremote,local,remote,ctimeout\fB)
is like
.B auth_tcpuser2
but
returns
NULL with errno set to ETIMEDOUT
if the authentication request has not been accepted or refused after
.I ctimeout
seconds.

.B auth_tcpuser4(\fIinlocal,inremote,local,remote,ctimeout,rtimeout\fB)
is like
.B auth_tcpuser3
but returns
NULL with errno set to ETIMEDOUT
if the reply to the authentication request is being sent slower than
.I rtimeout
seconds/character.

.B auth_fd(\fIfd,&inremote,&local,&remote\fB)
is the same as
.B auth_fd2
but throws away the
.I inlocal
information.
.B auth_tcpuser(\fIinremote,local,remote\fB)
is the same as
.B auth_tcpuser2
but may not bind to the proper local address on
hosts with multiple IP addresses.
These functions do not perform properly on multihomed
hosts and should not be used.
They are provided only for backwards compatibility.

The authentication routines check with the
remote Authentication Server on port
.B auth_tcpport,
which defaults to 113
as specified by RFC 931.
You can set
.B auth_tcpport
to other values
for nonstandard implementations.
.SH RESTRICTIONS
.B authuser
does no backslash interpretation
upon the remote user name.
This is conformance with the proposed revision to
RFC 931.

.B authuser
does not use the operating system type
information provided by the Authentication Server.
.SH VERSION
authuser version 4.0, February 9, 1992.
.SH AUTHOR
Placed into the public domain by Daniel J. Bernstein.
.SH REFERENCES
The authentication server is more secure than passwords
in some ways, but less secure than passwords in many ways.
(It's certainly better than no password at all---e.g., for
mail or news.)
It is not the final solution.
For an excellent discussion of security problems within
the TCP/IP protocol suite, see
Steve Bellovin's article
``Security Problems in the TCP/IP Protocol Suite.''
.SH "SEE ALSO"
tcpclient(1),
tcpserver(1),
getpeername(3),
getsockname(3),
tcp(4),
authd(8)


attic/src/libauth-4.0-p5/authuser.c
/*
 ** authuser.c
 **
 ** 2/2/93: Added auth_setreadtimeout() call.
 ** 7/21/92: Fixed SIGPIPE bug in auth_tcpuser3(). <pen@lysator.liu.se>
 ** 2/9/92: authuser 4.0. Public domain.
 ** 2/9/92: added bunches of zeroing just in case.
 ** 2/9/92: added auth_tcpuser3. uses bsd 4.3 select interface.
 ** 2/9/92: added auth_tcpsock, auth_sockuser.
 ** 2/9/92: added auth_fd2, auth_tcpuser2, simplified some of the code.
 ** 12/27/91: fixed up usercmp to deal with restricted tolower XXX
 ** 5/6/91 DJB baseline authuser 3.1. Public domain.
 */

#include <stdio.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/file.h>
#include <fcntl.h> /*XXX*/
#include <sys/socket.h>
#include <netinet/in.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
extern int errno;

#define _AUTHUSER_C
#include "authuser.h"

#ifndef FNDELAY
#  define FNDELAY O_NDELAY
#endif

unsigned short auth_tcpport = 113;
int            auth_rtimeout = 60;	/* 1 minute */

#define SIZ 500 /* various buffers */

static void clearsa(struct sockaddr_in *sa) {
    register char *x;
    for (x = (char *) sa;x < sizeof(*sa) + (char *) sa;++x)
	*x = 0;
}

static int usercmp( register char *u, register char *v) {
    register char uc = '\0';
    register char vc = '\0';
    register char ucvc = '\0';
    /* is it correct to consider Foo and fOo the same user? yes */
    /* but the function of this routine may change later */
    while ((uc = *u) && (vc = *v))
    {
	ucvc = (isupper(uc) ? tolower(uc) : uc) - (isupper(vc) ? tolower(vc) : vc);
	if (ucvc)
	    return ucvc;
	else
	    ++u,++v;
    }
    return uc || vc;
}

static char authline[SIZ];

char *auth_xline(register char *user, register int fd, register unsigned long *in)
{
    unsigned short local;
    unsigned short remote;
    register char *ruser;
    
    if (auth_fd(fd,in,&local,&remote) == -1)
	return 0;
    ruser = auth_tcpuser(*in,local,remote);
    if (!ruser)
	return 0;
    if (!user)
	user = ruser; /* forces X-Auth-User */
    sprintf(authline,
	    (usercmp(ruser,user) ? "X-Forgery-By: %s" : "X-Auth-User: %s"),
	    ruser);
    return authline;
}

int auth_fd( register int fd, register unsigned long *in,
    register unsigned short *local, register unsigned short *remote)
{
    unsigned long inlocal;
    return auth_fd2(fd,&inlocal,in,local,remote);
}

int auth_fd2( register int fd, register unsigned long *inlocal,
    register unsigned long *inremote, register unsigned short *local,
    register unsigned short *remote)
{
    struct sockaddr_in sa;
    int dummy;
    
    dummy = sizeof(sa);
    if (getsockname(fd,(struct sockaddr *)&sa,&dummy) == -1)
	return -1;
    if (sa.sin_family != AF_INET)
    {
	errno = EAFNOSUPPORT;
	return -1;
    }
    *local = ntohs(sa.sin_port);
    *inlocal = sa.sin_addr.s_addr;
    dummy = sizeof(sa);
    if (getpeername(fd,(struct sockaddr *)&sa,&dummy) == -1)
	return -1;
    *remote = ntohs(sa.sin_port);
    *inremote = sa.sin_addr.s_addr;
    return 0;
}

static char ruser[SIZ];
static char realbuf[SIZ];
static char *buf;

char *auth_tcpuser( register unsigned long in, register unsigned short local,
    register unsigned short remote)
{
    return auth_tcpuser2(0,in,local,remote);
}

#define CLORETS(e) { saveerrno = errno; close(s); errno = saveerrno; return e; }

int auth_tcpsock( register unsigned long inlocal,
    register unsigned long inremote)
{
    struct sockaddr_in sa;
    register int s;
    register int fl;
    register int saveerrno;
    
    if ((s = socket(AF_INET,SOCK_STREAM,0)) == -1)
	return -1;
    if (inlocal)
    {
	clearsa(&sa);
	sa.sin_family = AF_INET;
	sa.sin_port = 0;
	sa.sin_addr.s_addr = inlocal;
	if (bind(s,(struct sockaddr *)&sa,sizeof(sa)) == -1)
	    CLORETS(-1)
	    }
    if ((fl = fcntl(s,F_GETFL,0)) == -1)
	CLORETS(-1);
    if (fcntl(s,F_SETFL,FNDELAY | fl) == -1)
	CLORETS(-1);
    clearsa(&sa);
    sa.sin_family = AF_INET;
    sa.sin_port = htons(auth_tcpport);
    sa.sin_addr.s_addr = inremote;
    if (connect(s,(struct sockaddr *)&sa,sizeof(sa)) == -1)
	if (errno != EINPROGRESS)
	    CLORETS(-1)
		return s;
}

char *auth_tcpuser2( register unsigned long inlocal,
    register unsigned long inremote,
    register unsigned short local,
    register unsigned short remote)
{
    register int s;
    
    s = auth_tcpsock(inlocal,inremote);
    if (s == -1)
	return 0;
    return auth_sockuser(s,local,remote);
}

char *auth_tcpuser3( register unsigned long inlocal,
    register unsigned long inremote,
    register unsigned short local,
    register unsigned short remote,
    register int ctimeout)
{
    return auth_tcpuser4(inlocal,
			 inremote,
			 local,
			 remote,
			 ctimeout,
			 auth_rtimeout);
}


char *auth_tcpuser4( register unsigned long inlocal,
    register unsigned long inremote,
    register unsigned short local,
    register unsigned short remote,
    register int ctimeout,
    register int rtimeout)
{
    register int s;
    struct timeval ctv;
    fd_set wfds;
    register int r;
    register int saveerrno;
    void *old_sig;
    char *retval;
    
    
    old_sig = signal(SIGPIPE, SIG_IGN);
    
    s = auth_tcpsock(inlocal,inremote);
    if (s == -1)
    {
	signal(SIGPIPE, old_sig);
	return 0;
    }
    ctv.tv_sec = ctimeout;
    ctv.tv_usec = 0;
    FD_ZERO(&wfds);
    FD_SET(s,&wfds);
    r = select(s + 1,(fd_set *) 0,&wfds,(fd_set *) 0,&ctv);
    /* XXX: how to handle EINTR? */
    if (r == -1)
    {
	signal(SIGPIPE, old_sig);
	CLORETS(0);
    }
    if (!FD_ISSET(s,&wfds))
    {
	close(s);
	errno = ETIMEDOUT;
	signal(SIGPIPE, old_sig);
	return 0;
    }
    retval = auth_sockuser2(s,local,remote,rtimeout);
    signal(SIGPIPE, old_sig);
    
    return retval;
}

char *auth_sockuser( register int s,
    register unsigned short local,
    register unsigned short remote)
{
    return auth_sockuser2(s, local, remote, auth_rtimeout);
}

char *auth_sockuser2( register int s,
    register unsigned short local,
    register unsigned short remote,
    int rtimeout)
{
    register int buflen;
    register int w;
    register int saveerrno;
    char ch;
    unsigned short rlocal;
    unsigned short rremote;
    register int fl;
    fd_set wfds;
    void *old_sig;
    struct timeval rtv;
    
    old_sig = signal(SIGPIPE, SIG_IGN);
    
    FD_ZERO(&wfds);
    FD_SET(s,&wfds);
    
    select(s + 1,
	   (fd_set *) 0,
	   &wfds,(fd_set *) 0,
	   (struct timeval *) 0);
    
    /* now s is writable */
    if ((fl = fcntl(s,F_GETFL,0)) == -1)
    {
	signal(SIGPIPE, old_sig);
	CLORETS(0);
    }
    if (fcntl(s,F_SETFL,~FNDELAY & fl) == -1)
    {
	signal(SIGPIPE, old_sig);
	CLORETS(0);
    }
    buf = realbuf;
    sprintf(buf,"%u , %u\r\n",(unsigned int) remote,(unsigned int) local);
    /* note the reversed order---the example in RFC 931 is misleading */
    buflen = strlen(buf);
    while ((w = write(s,buf,buflen)) < buflen)
	if (w == -1) /* should we worry about 0 as well? */
	{
	    signal(SIGPIPE, old_sig);
	    CLORETS(0);
	}
	else
	{
	    buf += w;
	    buflen -= w;
	}
    buf = realbuf;
    
    do
    {
	fd_set rd_fds;
	
	rtv.tv_sec = rtimeout;
	rtv.tv_usec = 0;
    
	FD_ZERO(&rd_fds);
	FD_SET(s, &rd_fds);
	if (select(s+1,
		   &rd_fds,
		   (fd_set *) 0,
		   (fd_set *) 0,
		   &rtv) == 0)
	{
	    w = -1;
	    goto END;
	}
	
	if ((w = read(s,&ch,1)) == 1)
	{
	    *buf = ch;
	    if ((ch != ' ') && (ch != '\t') && (ch != '\r'))
		++buf;
	    if ((buf - realbuf == sizeof(realbuf) - 1) || (ch == '\n'))
		break;
	}
    } while (w == 1);
    
    END: 
    signal(SIGPIPE, old_sig);
    if (w == -1)
	CLORETS(0)
	    *buf = 0;
    
    if (sscanf(realbuf, "%hd,%hd: USERID :%*[^:]:%s",
	       &rremote, &rlocal, ruser) < 3)
    {
	close(s);
	errno = EIO;
	/* makes sense, right? well, not when USERID failed to match ERROR */
	/* but there's no good error to return in that case */
	return 0;
    }
    if ((remote != rremote) || (local != rlocal))
    {
	close(s);
	errno = EIO;
	return 0;
    }
    /* we're not going to do any backslash processing */
    close(s);
    return ruser;
}


